Add support for azure firewall

lonegunmanb opened the pull request:
## Describe your changes

## Issue number

#000

## Checklist before requesting a review
- [ ] The pr title can be used to describe what this pr did in `CHANGELOG.md` file
- [ ] I have executed pre-commit on my machine
- [ ] I have passed pr-check on my machine

Thanks for your cooperation!

matt-FFFFFF commented on main.tf:
```diff
@@ -90,7 +98,46 @@ resource "azurerm_subnet_route_table_association" "hub_routing_external" {
+  sku_name            = each.value.sku_name
+  sku_tier            = each.value.sku_tier
+  tags                = {}
+
```
I think we should support these to enable the functionality people will want:

- https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/firewall#private_ip_ranges
- https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/firewall#threat_intel_mode
- https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/firewall#zones
- https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/firewall#dns_servers
- https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/firewall#firewall_policy_id

lonegunmanb replied:
As for `firewall_policy_id`, since the resource group is created by us, I guess that the policy must be created by us too. I'll add this policy parameter as an embedded parameter inside firewall variable. @matt-FFFFFF

lonegunmanb replied:
I've changed my mind... Creating `azurerm_firewall_policy` would lead to the creation of user assigned identity, which requires the creation of resource group too. I think that'll make this module way too much complex. I just recall that we can create resource group outside this module, so I think we'd better leave user assigned identity / firewall policy out of this module. Just a simple firewall, if the user is not satisfied with that, then just create his own resource along with the firewall, firewall policy, or identity. WDYT?

matt-FFFFFF replied:
Hi @lonegunmanb 

In reality we probably don't want to be responsible for creating the firewall policy in this module.

A customer can link it to an existing one. One a policy is not supplied then we don't make the link.

Therefore we should remove the azurerm_firewall_policy from this module and leave it to be supplied by the caller.

matt-FFFFFF replied:
So, we create the firewall infrastructure as we need to know the internal IP address for routing purposes.

But, the firewall administration (the policy) is managed externally to this module.

matt-FFFFFF reviewed (changes requested):
lgtm - though we need a few more options exposed for the firewall. Feel free to ping me.

lonegunmanb commented:
Hi @matt-FFFFFF, I've added most of arguments you've mentioned, and added support for user-defined routes. It seems like we've finished all phrase one job, now we have created virtual networks, subnets, vnet peerings, the route tables and routes, Azure firewalls, and we've enabled our users to create their own route tables, routes and Firewalls or NVAs outside this module, please remind me if I've missed something.

If this pr looks good to you then I'll merge it and add some actual examples to verify it really works. Please let me know once you think we're good to go, thanks!
